### DISCUSSION TITLE
Pagination AND load more

### DISCUSSION THREAD

**Discussion by St-Permiakov** [vercel/swr#1546] [Q&A]
Hello!

I looked through issues, discussions and sandboxes, but couldn't find anything solid on using SWR for cases when you have simple pagination with pages and a "Load more" btn at the same time.
The usecase is: user can jump to, say, 6th page, and then load 7th and 8th, while keeping previous data.
Are there any examples for that case?

Which is better: standard useSWR or useSWRInfinite?

Am I right, that with useSWRInfinite if I want to load 9th page, for example, I'd have to do "setSize(9)" and load all 9 pages to show only one?
Or is it better to do just useSWR + react component state where I'd keep previous pages data if needed?

Thanks in advance!

**Comment by shuding** [comment]
Hi! Can you be more specific about the use case and behavior? I can't think of an app that has both infinite loading ("Load More" button), and random page accessing ("Jump to Page X").

**Reply by St-Permiakov** [reply]
Sure.
It's like pagination here:
https://www.mvideo.ru/stiralnye-i-sushilnye-mashiny-2427/stiralnye-mashiny-89/f/category=uzkie-stiralnye-mashiny-2446
You can see both pages and button with "Show more": http://joxi.ru/5md9y5oHaLoxJ2

That's exactly the behavior I need.

**Reply by michalpulpan** [reply]
I've seen it in several ecommerce store fronts as well. It's quite handy to have this feature. 
If there's any example of this implementation, I'd be interested in seeing that as well.
Thanks

**Comment by michalpulpan** [comment]
@St-Permiakov DId you figgure it out? I, for now, changed my page to just use "Load more", but would love to have an option to just change to `?page=4` for example.

**Reply by St-Permiakov** [reply]
Not yet. Postponed this issue a bit, but soon will try again.
Sadly I can't drop part of the functionality, it's not up to me to decide.

**Comment by St-Permiakov** [comment]
The question still stands, any help will be appreciated!

**Reply by michalpulpan** [reply]
@St-Permiakov I ended up writing my own logic which handles this issue using standard SWR and simple state.

It's not the most effective solution, but it works on quite large data so the performance is pretty okay. Or at least - it's not a bottle neck.
It's not really a copy-paste solution, but it's a simple idea to implement.

```javascript

const CategoryPage = ({
        // some parameters
	}: ICategoryPage) => {
              

               const [products, setProducts] = useState([]); // State holding array of products
               const [activePage, setActivePage] = useState < number > (queryParamsState?.page || 1); // I have one large reducer object holding all products filters, ordering, and also page which is synchronized with user URL, so I just read the page parameter
              const [loadMore, setLoadMore] = useState < boolean > (false);
             
              const {
			data: clientProducts,
			error: clientProductsError,
			mutate: clientProductsMutate,
		} = useSWR(`/category/${id}/products?page=${activePage}`); 
             
              

               

              

 
               // If SWR fetches new data
               // if user uses LoadMore button, it append into the products array
               // if user uses Pagination button (else branch), just replace the array with newly fetched data
		useEffect(()=>{
			if(clientProducts === undefined || !clientProducts && clientProducts.length === 0){
			} else if(loadMore) {
				setProducts(prevProducts => [...prevProducts, clientProducts])
				setLoadMore(false);
			} else {
				setProducts([clientProducts])
			}
		}, [clientProducts])


                // and in return just simply render some "ProductList" component which will iterate over array of products
                // and create pagination component with numbers, left/right arrow and LoadMore button

                return (
                     {/* PRODUCT LIST */}
                     ...
                     {/* PRODUCT LIST */}

                     <Pagination
				pageCount={products[0]?.total_pages}
				currentPage={activePage}
				setter={setDispatcher}
				showMoreButton={products[0]?.total_pages > activePage}
				showMoreButtonSetter={() => {
				        setActivePage(activePage + 1);
				        setLoadMore(true);
					setDispatcher('page', activePage + 1);
					clientProductsMutate();
				}}
			/>
                  
               )
}

/* And just create pagination component which will generate page buttons, and load more button which will use showMoreButtonSetter parameter 
   And of course it's needed to somehow handle setter for page buttons
 */

```

Hope it helps at least with some kind of an approach 😊.

**Reply by St-Permiakov** [reply]
Thank you!
Same idea came to my mind, and ended up with a similar approach.
I made a hook that memorizes previous data if needed:
```ts
// 📂 hook
export const usePaginationWithMemory = <T>(startPage = 1, startData: T[] = []) => {
  const [page, setPage] = useState(startPage);
  const [prevData, setPrevData] = useState(startData);

  const setNewPage = (newPage: number, dataToKeep?: any[]) => {
    setPage(newPage);
    setPrevData(dataToKeep ? prevState => ([...prevState, ...dataToKeep]) : []);
  };

  // helper to resolve conflict between "laggy" SWR middleware and this hook
  const cleanData = (data) => data.filter((item, index) => data.indexOf(item) === index);

  return { page, prevData, setNewPage, cleanData };
};

// 📂 Component with pagination
{
    ...
    const { page, setNewPage, prevData, cleanData } = usePaginationWithMemory<TProduct>(startPage);

    // SWR request
    const { products } = useProducts(page);

    // change page event
    const handleChangePage = (page: number, more: boolean) => {
      setNewPage(page, more ? products : undefined);
    }

    const data = cleanData([...prevData, ...products]);
}
```

It can be easily applied to any component with pagination and make less rerenders (only 2 actually).

The only thing I stumbled upon was conflict between "laggy" SWR (when we want previous data to be shown until new data is loaded) and my implementation. While loading it showed both my "memorized" items and those kept by SWR "laggy" behavior.
This leads to React errors like "Encountered two children with the same key" and strange visual clones.
I "fixed" it by filtering data, but it's neither elegant nor effective decision.

I hope I'll figure out how to combine those ideas in a performant and reusable way.
